### ISSUE TITLE
Integrate with credential-storages

### ISSUE THREAD

**Issue by xakep666** [deviceinsight/kafkactl#133]
Currently kafkactl stores passwords (for sasl or maybe tls certs in the future) inside config file. It is not secure because passwords available for any application running from current user.
I want to discuss about integration with special credential storage. I.e. Docker uses https://github.com/docker/docker-credential-helpers to interact with keychain, wincred, etc.
I propose to use https://github.com/99designs/keyring library because it has integration with many system storages and truly cross-platform encrypted-file based storage. But kafkactl built with disabled cgo but keychain backend requires it. User can specify global storage backends settings in config. When something requires passwords kafkactl attempts to find it by "context key" (context name+auth method type or certificate fingerprint). If credentials was not found user will be prompted for them. Credentials may be reset with commandline flag.
Any thoughs?

**Comment by d-rk** [comment]
Hi @xakep666,

generally I would appreciate this to be implemented.
But I don't want to sacrifice disabled cgo for such a small thing, because enabling cgo will probably lead to more problems when building for the different OSes/architectures.

the need for cgo in https://github.com/99designs/keyring is only for mac right? So one could live without mac support for now.
There is also https://github.com/zalando/go-keyring which supports fewer backends, but seems to work without cgo enabled.

The best option would probably be to contribute a keychain backend that does not require cgo to  **99designs/keyring**.
But I don't use Mac, so at least for me thats not an option.

**Comment by d-rk** [comment]
this should also cover https://github.com/deviceinsight/kafkactl/issues/133

**Comment by runelan** [comment]
We have a case related to the fact that passwords are kept in files. Imho adding a prompt that asks the user for their password would be slightly better in this regard. Any thoughts on this, or would you prefer that discussion for a separate issue?

**Comment by d-rk** [comment]
Thats probably most of the effort to implement this here.

Asking for passwords currently affects:
- tls.certKey
- schemaRegistry.tls.certKey
- sasl.password
- schemaRegistry.password

For the password fields, we can probably ask for password if username field is set, but the password is missing.

For the certKey we need currently dont support passphrases. We should add a config key certKeyPassphrase. Then we somehow need to detect if the certKey is passphrase protected, if yes use new certKeyPassphrase if available, otherwise ask.

For the keyring integration i still favor https://github.com/zalando/go-keyring. Integrating it would just require to query keyring before asking for a password and saving to keyring after asking. Keyring usage should be enabled per default, but one must be able to disable usage via config flag.

**Comment by xakep666** [comment]
I would suggest to _not_ use zalando keyring because of https://github.com/zalando/go-keyring/issues/110
TLDR: for macos keychain ownership of item is tied to calling application which is in that case is a globally available /usr/bin/security
Here is PR https://github.com/zalando/go-keyring/pull/136 that changes `exec` based integration to `purego` based direct calls to security framework (keychain is part of it)

**Comment by d-rk** [comment]
Ok, thanks for pointing that out.

**Comment by d-rk** [comment]
@xakep666 I decided to go with zalando/go-keyring nevertheless. It is better than what we had so far.
Also the keyring implementation is quite isolated and easy to replace, but  from the comments in https://github.com/zalando/go-keyring/pull/136 I'm optimistic that it will improve in the future.
